README.md
# NAME

MooseX::Declare - Declarative syntax for Moose

# SYNOPSIS

    use MooseX::Declare;

    class BankAccount {
        has 'balance' => ( isa => 'Num', is => 'rw', default => 0 );

        method deposit (Num $amount) {
            $self->balance( $self->balance + $amount );
        }

        method withdraw (Num $amount) {
            my $current_balance = $self->balance();
            ( $current_balance >= $amount )
                || confess "Account overdrawn";
            $self->balance( $current_balance - $amount );
        }
    }

    class CheckingAccount extends BankAccount {
        has 'overdraft_account' => ( isa => 'BankAccount', is => 'rw' );

        before withdraw (Num $amount) {
            my $overdraft_amount = $amount - $self->balance();
            if ( $self->overdraft_account && $overdraft_amount > 0 ) {
                $self->overdraft_account->withdraw($overdraft_amount);
                $self->deposit($overdraft_amount);
            }
        }
    }

# DESCRIPTION

This module provides syntactic sugar for Moose, the postmodern object system
for Perl 5. When used, it sets up the `class` and `role` keywords.

__Note:__ Please see the ["WARNING"](#WARNING) section below!

# KEYWORDS

## class

    class Foo { ... }

    my $anon_class = class { ... };

Declares a new class. The class can be either named or anonymous, depending on
whether or not a classname is given. Within the class definition Moose and
[MooseX::Method::Signatures](http://search.cpan.org/perldoc?MooseX::Method::Signatures) are set up automatically in addition to the other
keywords described in this document. At the end of the definition the class
will be made immutable. namespace::autoclean is injected to clean up Moose and
other imports for you.

Because of the way the options are parsed, you cannot have a class named "is",
"with" or "extends".

It's possible to specify options for classes:

- extends

        class Foo extends Bar { ... }

    Sets a superclass for the class being declared.

- with

        class Foo with Role             { ... }
        class Foo with Role1 with Role2 { ... }
        class Foo with (Role1, Role2)   { ... }

    Applies a role or roles to the class being declared.

- is mutable

        class Foo is mutable { ... }

    Causes the class not to be made immutable after its definition.

    Options can also be provided for anonymous classes using the same syntax:

        my $meta_class = class with Role;

## role

    role Foo { ... }

    my $anon_role = role { ... };

Declares a new role. The role can be either named or anonymous, depending on
whether or not a name is given. Within the role definition Moose::Role and
MooseX::Method::Signatures are set up automatically in addition to the other
keywords described in this document. Again, namespace::autoclean is injected to
clean up Moose::Role and other imports for you.

It's possible to specify options for roles:

- with

        role Foo with Bar { ... }

    Applies a role to the role being declared.

## before / after / around / override / augment

    before   foo ($x, $y, $z) { ... }
    after    bar ($x, $y, $z) { ... }
    around   baz ($x, $y, $z) { ... }
    override moo ($x, $y, $z) { ... }
    augment  kuh ($x, $y, $z) { ... }

Add a method modifier. Those work like documented in [Moose](http://search.cpan.org/perldoc?Moose), except for
the slightly nicer syntax and the method signatures, which work like documented
in [MooseX::Method::Signatures](http://search.cpan.org/perldoc?MooseX::Method::Signatures).

For the `around` modifier an additional argument called `$orig` is
automatically set up as the invocant for the method.

## clean

Sometimes you don't want the automatic cleaning the `class` and `role`
keywords provide using namespace::autoclean. In those cases you can specify the
`dirty` trait for your class or role:

    use MooseX::Declare;
    class Foo is dirty { ... }

This will prevent cleaning of your namespace, except for the keywords imported
from `Moose` or `Moose::Role`. Additionally, a `clean` keyword is provided,
which allows you to explicitly clean all functions that were defined prior to
calling `clean`. Here's an example:

    use MooseX::Declare;
    class Foo is dirty {
        sub helper_function { ... }
        clean;
        method foo ($stuff) { ...; return helper_function($stuff); }
    }

With that, the helper function won't be available as a method to a user of your
class, but you're still able to use it inside your class.

# NOTE ON IMPORTS

When creating a class with MooseX::Declare like:

    use MooseX::Declare;
    class Foo { ... }

What actually happens is something like this:

    {
        package Foo;
        use Moose;
        use namespace::autoclean;
        ...
        __PACKAGE__->meta->make_immutable;
    }

So if you declare imports outside the class, the symbols get imported into the
`main::` namespace, not the class' namespace. The symbols then cannot be called
from within the class:

    use MooseX::Declare;
    use Data::Dump qw/dump/;
    class Foo {
        method dump($value) { return dump($value) } # Data::Dump::dump IS NOT in Foo::
        method pp($value)   { $self->dump($value) } # an alias for our dump method
    }

To solve this, only import MooseX::Declare outside the class definition
(because you have to). Make all other imports inside the class definition.

    use MooseX::Declare;
    class Foo {
        use Data::Dump qw/dump/;
        method dump($value) { return dump($value) } # Data::Dump::dump IS in Foo::
        method pp($value)   { $self->dump($value) } # an alias for our dump method
    }

    Foo->new->dump($some_value);
    Foo->new->pp($some_value);

__NOTE__ that the import `Data::Dump::dump()` and the method `Foo::dump()`,
although having the same name, do not conflict with each other, because the
imported `dump` function will be cleaned during compile time, so only the
method remains there at run time. If you want to do more esoteric things with
imports, have a look at the `clean` keyword and the `dirty` trait.

# WARNING

__Warning:__ MooseX::Declare is based on [Devel::Declare](http://search.cpan.org/perldoc?Devel::Declare), a giant bag of crack
originally implemented by mst with the goal of upsetting the perl core
developers so much by its very existence that they implemented proper
keyword handling in the core.

As of perl5 version 14, this goal has been achieved, and modules such
as [Devel::CallParser](http://search.cpan.org/perldoc?Devel::CallParser), [Function::Parameters](http://search.cpan.org/perldoc?Function::Parameters), and [Keyword::Simple](http://search.cpan.org/perldoc?Keyword::Simple) provide
mechanisms to mangle perl syntax that don't require hallucinogenic
drugs to interpret the error messages they produce.

If you want to use declarative syntax in new code, please for the love
of kittens get yourself a recent perl and look at [Moops](http://search.cpan.org/perldoc?Moops) instead.

# SEE ALSO

- [Moose](http://search.cpan.org/perldoc?Moose)
- [Moose::Role](http://search.cpan.org/perldoc?Moose::Role)
- [MooseX::Method::Signatures](http://search.cpan.org/perldoc?MooseX::Method::Signatures)
- [namespace::autoclean](http://search.cpan.org/perldoc?namespace::autoclean)
- vim syntax: [http://www.vim.org/scripts/script.php?script\_id=2526](http://www.vim.org/scripts/script.php?script\_id=2526)
- emacs syntax: [http://github.com/jrockway/cperl-mode](http://github.com/jrockway/cperl-mode)
- Geany syntax + notes: [http://www.cattlegrid.info/blog/2009/09/moosex-declare-geany-syntax.html](http://www.cattlegrid.info/blog/2009/09/moosex-declare-geany-syntax.html)
- [Devel::CallParser](http://search.cpan.org/perldoc?Devel::CallParser)
- [Function::Parameters](http://search.cpan.org/perldoc?Function::Parameters)
- [Keyword::Simple](http://search.cpan.org/perldoc?Keyword::Simple)
- [Moops](http://search.cpan.org/perldoc?Moops)

# AUTHOR

Florian Ragwitz <rafl@debian.org>

# CONTRIBUTORS

- Ash Berlin <ash\_github@firemirror.com>
- Chas. J. Owens IV <chas.owens@gmail.com>
- Chris Prather <chris@prather.org>
- Dave Rolsky <autarch@urth.org>
- Devin Austin <dhoss@cpan.org>
- Hans Dieter Pearcey <hdp@cpan.org>
- Justin Hunter <justin.d.hunter@gmail.com>
- Karen Etheridge <ether@cpan.org>
- Matt Kraai <kraai@ftbfs.org>
- Michele Beltrame <arthas@cpan.org>
- Nelo Onyiah <nelo.onyiah@gmail.com>
- Nick Perez <nperez@cpan.org>
- Piers Cawley <pdcawley@bofh.org.uk>
- Rafael Kitover <rkitover@io.com>
- Robert 'phaylon' Sedlacek <rs@474.at>
- Stevan Little <stevan.little@iinteractive.com>
- Tomas Doran <bobtfish@bobtfish.net>
- Yanick Champoux <yanick@babyl.dyndns.org>
- leedo <lee@laylward.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2008 by Florian Ragwitz.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for MooseX-Declare

{{$NEXT}}

0.43      2015-08-16 03:29:28Z
  - add some diagnostics to a test that is failing occasionally

0.42      2015-06-04 06:17:43Z
  - eliminate remaining uses of List::MoreUtils
  - increase required version of namespace::autoclean to 0.16 to avoid test
    failures

0.41      2015-05-16 03:35:49Z
  - mark this distribution as deprecated in metadata

0.40      2014-11-12 02:27:58Z
  - mention deprecation status in module ABSTRACT
  - fix warning in tests on 5.8.x

0.39      2014-08-16 19:12:31Z
  - line numbers in shipped code are now almost the same (within 1) as
    the repository source, for easier debugging
  - pod fixes (Frank Wiegand, RT#51722)

0.38      2013-12-14 18:39:30Z
  - update configure_requires checking in Makefile.PL, add CONTRIBUTING file

0.37      2013-09-26 04:08:21Z
  - repository migrated to the github moose organization
  - re-release with fixed compile test

0.36      2013-09-10 02:33:28Z
  - Depend on 'aliased'.
  - The test suite now uses Test::Fatal instead of Test::Exception. (Karen
    Etheridge)
  - removed use of deprecated Class::MOP::load_class, calling enum with a list
  - converted use of Class::MOP::class_of to Moose::Util::find_meta
  - new WARNING section added to pod, regarding future utility of this module

0.35   2011-08-23 11:13:10Z
  - Adapt to changes in MooseX::Role::Parameterized 0.27.

0.34   2010-08-31 23:10:28Z
  * Depend on MooseX::Method::Signatures 0.36 to avoid deprecation warnings on
    new versions of Moose (Jeremy Carman).

0.33   2010-02-04 01:50:14Z
  - Depend on MooseX::Method::Signatures 0.30 to fix various incompatibilities.

0.32   2009-11-02 02:00:42Z
  - Add link to Geany syntax notes to documentation (Michele Beltrame).
  - Convert from MooseX::MethodAttributes to Native::Trait (Justin Hunter).
  - Update dependency on MooseX::Types (Chris Prather).

0.31   2009-09-17 16:10:22Z
  - Bump dependencies for no more warnings on new Moose versions.

0.30   2009-09-05 00:48:05Z
  - Make the attr inheriting TODO test work.
  - Make default values for role parameters work (nperez).
  - Add tests for the above (nperez).

0.29   2009-08-31 17:11:25Z
  - Add failing test for applying modifiers to a method directly composed from
    a role (nperez).
  - Fix the failing test by depending on MXMS 0.24.

0.28   2009-08-27 20:44:34Z
  - Bump prereqs on MooseX::Method::Signatures, Moose, and namespace::clean in
    order to avoid test failure.

0.27   2009-08-18 06:54:22Z
  - Depend on MooseX::Method::Signatures 0.21 to make sure type aliasing works
    and there are no warnings due to the recent MooseX::LazyRequire changes.

0.26   2009-08-16 12:48:12Z
  - Declare dependency on MooseX::AttributeHelpers.

0.25   2009-08-16 03:31:59Z
  - Don't persist the inner keywords over more than one invocation of the declarator.
  - Pass along the current context to import_symbols_from and imported_moose_symbols.
  - Implement parameterizable roles.

0.24   2009-08-15 21:48:09Z
  - Add repository and bugtracker to Makefile.PL (Ash Berlin).
  - Tests and doc fixes on applying multiple roles at once (Chas. J. Owens IV).
  - Produce better diagnostic output on test failures (Dave Rolsky).
  - Make MooseX::Method::Signatures handle the actual parsing for method-like
    keywords (nperez).
  - Clean now warns in autoclean namespaces (Robert 'phaylon' Sedlacek).
  - 'method' keyword now only available by default in namespaces that consumed
    MooseSetup (Robert 'phaylon' Sedlacek).
  - Added 'namespace' keyword and allow namespaces, superclasses and roles to
    be relative (Robert 'phaylon' Sedlacek).
  - Add tests for using anonymous methods as attribute defaults (Devin Austin).
  - Stop depending on B::Hooks::EndOfScope. We don't use it direectly anymore.
  - Add a real 'with' keyword.
  - 'with' in a class- or role-body doesn't defer role application until the
    end of the class- or role-definition anymore. The 'with'-option for the
    class and role keywords still does.
  - Add a tests for role application and method modifier ordering.
  - Make anon classes immutable as well (unless explicitly asked not to).
  - Add a ctx method to add cleanup code that should be injected as early as
    possible.
  - Depend on recent versions of DD and MXMS to make keywords less reserved
    (i.e. hash keys and barewords before fat commas).
  - Make t/with_newlines.t less noisy.
  - Add tests for methods with no signature not caring about their arguments,
    other than $self.

0.23   2009-06-25 03:09:10Z
  - Add test to make sure methods returned from meta are
    MooseX::Method::Signatures::Meta::Method (nperez).
  - Add the actual ::Signature methods to the meta class instead of letting it
    wrap the coderef itself.
  - Make inner scopes inherit the declaration provider instead of hardcoding
    'MooseX::Declare' (Robert 'phaylon' Sedlacek).
  - Fix test failure on new Moose versions, which inline a DESTROY method on
    immutalisation.

0.22   2009-05-06 00:37:37Z
  - Add TODO tests for creating nested anonymous classes and roles
    (nperez).
  - Always end injected code with a semicolon.
  - Only do namespace nesting if the inner namespace starts with "::".
  - Depend on D::D 0.005001 to get a skip_declarator that doesn't fail
    when code is being compiled between calling the linstr callback and
    skip_declarator. This appears to be necessary to work on 5.8.

0.21   2009-05-04 23:55:41Z
  - Add tests for automatic cleaning with the clean/dirty traits
    (Stevan Little).
  - Automatically clean all imports at the end of compile time unless the
    dirty trait is given.
  - Clean now warns in autoclean namespaces (Robert Sedlacek).
  - Document all refactored components (Robert Sedlacek).
  - Improve error reporting (Robert Sedlacek).
  - Added link to emacs syntax highlighter (Nelo Onyiah).
  - Fix a doc typo (reported by Alexander Hartmaier).

lib/MooseX/Declare.pm
use strict;
use warnings;

package MooseX::Declare;
# ABSTRACT: (DEPRECATED) Declarative syntax for Moose
# KEYWORDS: moose extension declaration syntax sugar method class deprecated

our $VERSION = '0.44';

use aliased 'MooseX::Declare::Syntax::Keyword::Class',      'ClassKeyword';
use aliased 'MooseX::Declare::Syntax::Keyword::Role',       'RoleKeyword';
use aliased 'MooseX::Declare::Syntax::Keyword::Namespace',  'NamespaceKeyword';

use namespace::clean 0.19;

sub import {
    my ($class, %args) = @_;

    my $caller = caller();

    strict->import;
    warnings->import;

    for my $keyword ($class->keywords) {
        $keyword->setup_for($caller, %args, provided_by => $class);
    }
}

sub keywords {
    ClassKeyword->new(identifier => 'class'),
    RoleKeyword->new(identifier => 'role'),
    NamespaceKeyword->new(identifier => 'namespace'),
}

=head1 SYNOPSIS

    use MooseX::Declare;

    class BankAccount {
        has 'balance' => ( isa => 'Num', is => 'rw', default => 0 );

        method deposit (Num $amount) {
            $self->balance( $self->balance + $amount );
        }

        method withdraw (Num $amount) {
            my $current_balance = $self->balance();
            ( $current_balance >= $amount )
                || confess "Account overdrawn";
            $self->balance( $current_balance - $amount );
        }
    }

    class CheckingAccount extends BankAccount {
        has 'overdraft_account' => ( isa => 'BankAccount', is => 'rw' );

        before withdraw (Num $amount) {
            my $overdraft_amount = $amount - $self->balance();
            if ( $self->overdraft_account && $overdraft_amount > 0 ) {
                $self->overdraft_account->withdraw($overdraft_amount);
                $self->deposit($overdraft_amount);
            }
        }
    }

=head1 DESCRIPTION

This module provides syntactic sugar for Moose, the postmodern object system
for Perl 5. When used, it sets up the C<class> and C<role> keywords.

B<Note:> Please see the L</WARNING> section below!

=head1 KEYWORDS

=head2 class

    class Foo { ... }

    my $anon_class = class { ... };

Declares a new class. The class can be either named or anonymous, depending on
whether or not a classname is given. Within the class definition Moose and
L<MooseX::Method::Signatures> are set up automatically in addition to the other
keywords described in this document. At the end of the definition the class
will be made immutable. namespace::autoclean is injected to clean up Moose and
other imports for you.

Because of the way the options are parsed, you cannot have a class named "is",
"with" or "extends".

It's possible to specify options for classes:

=over 4

=item extends

    class Foo extends Bar { ... }

Sets a superclass for the class being declared.

=item with

    class Foo with Role             { ... }
    class Foo with Role1 with Role2 { ... }
    class Foo with (Role1, Role2)   { ... }

Applies a role or roles to the class being declared.

=item is mutable

    class Foo is mutable { ... }

Causes the class not to be made immutable after its definition.

Options can also be provided for anonymous classes using the same syntax:

    my $meta_class = class with Role;

=back

=head2 role

    role Foo { ... }

    my $anon_role = role { ... };

Declares a new role. The role can be either named or anonymous, depending on
whether or not a name is given. Within the role definition Moose::Role and
MooseX::Method::Signatures are set up automatically in addition to the other
keywords described in this document. Again, namespace::autoclean is injected to
clean up Moose::Role and other imports for you.

It's possible to specify options for roles:

=over 4

=item with

    role Foo with Bar { ... }

Applies a role to the role being declared.

=back

=head2 before / after / around / override / augment

    before   foo ($x, $y, $z) { ... }
    after    bar ($x, $y, $z) { ... }
    around   baz ($x, $y, $z) { ... }
    override moo ($x, $y, $z) { ... }
    augment  kuh ($x, $y, $z) { ... }

Add a method modifier. Those work like documented in L<Moose|Moose>, except for
the slightly nicer syntax and the method signatures, which work like documented
in L<MooseX::Method::Signatures|MooseX::Method::Signatures>.

For the C<around> modifier an additional argument called C<$orig> is
automatically set up as the invocant for the method.

=head2 clean

Sometimes you don't want the automatic cleaning the C<class> and C<role>
keywords provide using namespace::autoclean. In those cases you can specify the
C<dirty> trait for your class or role:

    use MooseX::Declare;
    class Foo is dirty { ... }

This will prevent cleaning of your namespace, except for the keywords imported
from C<Moose> or C<Moose::Role>. Additionally, a C<clean> keyword is provided,
which allows you to explicitly clean all functions that were defined prior to
calling C<clean>. Here's an example:

    use MooseX::Declare;
    class Foo is dirty {
        sub helper_function { ... }
        clean;
        method foo ($stuff) { ...; return helper_function($stuff); }
    }

With that, the helper function won't be available as a method to a user of your
class, but you're still able to use it inside your class.

=head1 NOTE ON IMPORTS

When creating a class with MooseX::Declare like:

    use MooseX::Declare;
    class Foo { ... }

What actually happens is something like this:

    {
        package Foo;
        use Moose;
        use namespace::autoclean;
        ...
        __PACKAGE__->meta->make_immutable;
    }

So if you declare imports outside the class, the symbols get imported into the
C<main::> namespace, not the class' namespace. The symbols then cannot be called
from within the class:

    use MooseX::Declare;
    use Data::Dump qw/dump/;
    class Foo {
        method dump($value) { return dump($value) } # Data::Dump::dump IS NOT in Foo::
        method pp($value)   { $self->dump($value) } # an alias for our dump method
    }

To solve this, only import MooseX::Declare outside the class definition
(because you have to). Make all other imports inside the class definition.

    use MooseX::Declare;
    class Foo {
        use Data::Dump qw/dump/;
        method dump($value) { return dump($value) } # Data::Dump::dump IS in Foo::
        method pp($value)   { $self->dump($value) } # an alias for our dump method
    }

    Foo->new->dump($some_value);
    Foo->new->pp($some_value);

B<NOTE> that the import C<Data::Dump::dump()> and the method C<Foo::dump()>,
although having the same name, do not conflict with each other, because the
imported C<dump> function will be cleaned during compile time, so only the
method remains there at run time. If you want to do more esoteric things with
imports, have a look at the C<clean> keyword and the C<dirty> trait.

=head1 WARNING

=for comment rafl agreed we should have a warning, and mst wrote this:

B<Warning:> MooseX::Declare is based on L<Devel::Declare>, a giant bag of crack
originally implemented by mst with the goal of upsetting the perl core
developers so much by its very existence that they implemented proper
keyword handling in the core.

As of perl5 version 14, this goal has been achieved, and modules such
as L<Devel::CallParser>, L<Function::Parameters>, and L<Keyword::Simple> provide
mechanisms to mangle perl syntax that don't require hallucinogenic
drugs to interpret the error messages they produce.

If you want to use declarative syntax in new code, please for the love
of kittens get yourself a recent perl and look at L<Moops> instead.

=head1 SEE ALSO

=for :list
* L<Moose>
* L<Moose::Role>
* L<MooseX::Method::Signatures>
* L<namespace::autoclean>
* vim syntax: L<http://www.vim.org/scripts/script.php?script_id=2526>
* emacs syntax: L<http://github.com/jrockway/cperl-mode>
* Geany syntax + notes: L<http://www.cattlegrid.info/blog/2009/09/moosex-declare-geany-syntax.html>
* L<Devel::CallParser>
* L<Function::Parameters>
* L<Keyword::Simple>
* L<Moops>

=cut


1;
